src/python/pants/engine/fs_test.py

```diff
@@ -40,871 +37,711 @@
+            QueryRule(Snapshot, [CreateDigest]),
+            QueryRule(Snapshot, [DigestSubset]),
+            QueryRule(Snapshot, [PathGlobs]),
+        ],
+        isolated_local_store=True,
+    )
 
-    def prime_store_with_roland_digest(self) -> Digest:
-        """This method primes the store with a directory of a file named 'roland' and contents
-        'European Burmese'."""
-        with temporary_dir() as temp_dir:
-            with open(os.path.join(temp_dir, "roland"), "w") as f:
-                f.write("European Burmese")
-            globs = PathGlobs(["*"])
-            snapshot = self.scheduler.capture_snapshots((PathGlobsAndRoot(globs, temp_dir),))[0]
 
-            expected_digest = Digest(
-                "63949aa823baf765eff07b946050d76ec0033144c785a94d3ebd82baa931cd16", 80
-            )
-            self.assert_snapshot_equals(snapshot, ["roland"], expected_digest)
-        return expected_digest
+ROLAND_DIGEST = Digest("63949aa823baf765eff07b946050d76ec0033144c785a94d3ebd82baa931cd16", 80)
+
+
+def prime_store_with_roland_digest(rule_runner: RuleRunner) -> None:
+    """Prime lmdb_store with a directory of a file named 'roland' and contents 'European
+    Burmese'."""
+    with temporary_dir() as temp_dir:
+        Path(temp_dir, "roland").write_text("European Burmese")
+        snapshot = rule_runner.scheduler.capture_snapshots(
+            (PathGlobsAndRoot(PathGlobs(["*"]), temp_dir),)
+        )[0]
+    assert snapshot.files == ("roland",)
+    assert snapshot.digest == ROLAND_DIGEST
+
+
+def setup_fs_test_tar(rule_runner: RuleRunner) -> None:
+    """Extract fs_test.tar into the rule_runner's build root.
+
+    Note that we use a tar, rather than rule_runner.create_file(), because it has symlinks set up a
+    certain way.
+
+    Contents:
+
+        4.txt
+        a
+        ├── 3.txt
+        ├── 4.txt.ln -> ../4.txt
+        └── b
+            ├── 1.txt
+            └── 2
+        c.ln -> a/b
+        d.ln -> a
+    """
+    data = pkgutil.get_data("pants.engine.internals", "fs_test_data/fs_test.tar")
+    assert data is not None
+    io = BytesIO()
+    io.write(data)
+    io.seek(0)
+    with tarfile.open(fileobj=io) as tf:
+        tf.extractall(rule_runner.build_root)
+
+
+FS_TAR_ALL_FILES = (
+    "4.txt",
+    "a/3.txt",
+    "a/4.txt.ln",
+    "a/b/1.txt",
+    "a/b/2",
+    "c.ln/1.txt",
+    "c.ln/2",
+    "d.ln/3.txt",
+    "d.ln/4.txt.ln",
+    "d.ln/b/1.txt",
+    "d.ln/b/2",
+)
+FS_TAR_ALL_DIRS = ("a", "a/b", "c.ln", "d.ln", "d.ln/b")
+
+
+def try_with_backoff(assertion_fn: Callable[[], bool]) -> bool:
+    for i in range(4):
+        time.sleep(0.1 * i)
+        if assertion_fn():
+            return True
+    return False
+
+
+# -----------------------------------------------------------------------------------------------
+# `PathGlobs`, including `GlobMatchErrorBehavior` and symlink handling
+# -----------------------------------------------------------------------------------------------
```

Comment:
These tests had the most refactoring, as I removed several helper functions like `assert_walks_files` and we don't use `ProjectTree` anymore.

I consolidated a couple tests into a single test. I was careful to make sure we're still testing the same things.